File: python/shared/dtos/auction.py
from enum import IntEnum
from typing import Any

from pydantic import Field

from . import BaseDTO, BigInt
from .preset import PresetDetailDTO


AUCTION_LIFETIME = 3600


class Status(IntEnum):
    WAITING = 0
    PENDING = 1
    RUNNING = 2
    COMPLETED = 3


class AuctionClientEventType(IntEnum):
    AUTH = 0
    PLACE_BID = 1


class AuctionPublishType(IntEnum):
    TICK = 2
    STATUS = 3
    BID_PLACED = 4
    MEMBER_SOLD = 5
    MEMBER_UNSOLD = 6
    LEADER_CONNECTED = 7
    LEADER_DISCONNECTED = 8
    NEXT_PLAYER = 9


class AuctionServerEventType(IntEnum):
    INIT = 0
    ERROR = 1
    TICK = 2
    STATUS = 3
    BID_PLACED = 4
    MEMBER_SOLD = 5
    MEMBER_UNSOLD = 6
    LEADER_CONNECTED = 7
    LEADER_DISCONNECTED = 8
    NEXT_PLAYER = 9


class AuctionRequestType(IntEnum):
    CREATE = 0
    PLACE_BID = 1
    LEADER_CONNECTED = 2
    LEADER_DISCONNECTED = 3
    RECOVER = 4


class AuctionResponseType(IntEnum):
    BID_ERROR = 0


class TeamDTO(BaseDTO):
    team_id: int
    leader_id: int
    member_ids: list[int]
    points: int


class BidDTO(BaseDTO):
    amount: int
    leader_id: int


class AuctionDTO(BaseDTO):
    auction_id: BigInt


class AuctionDetailDTO(AuctionDTO):
    status: Status
    connected_leader_count: int
    player_id: int | None
    bid: BidDTO | None
    teams: list[TeamDTO]
    auction_queue: list[int]
    unsold_queue: list[int]
    preset_snapshot: PresetDetailDTO | None
    ttl: int
    timer: int


class InitEventPayloadDTO(BaseDTO):
    auction: AuctionDetailDTO
    member_id: int | None


class PlaceBidEventPayloadDTO(BaseDTO):
    amount: int = Field(ge=1, le=10000)


class AuthEventPayloadDTO(BaseDTO):
    access_token: str | None


class ErrorEventPayloadDTO(BaseDTO):
    code: int


class TickEventPayloadDTO(BaseDTO):
    timer: int


class StatusEventPayloadDTO(BaseDTO):
    status: Status


class BidPlacedEventPayloadDTO(BaseDTO):
    leader_id: int
    amount: int


class AuctionClientEventEnvelopeDTO(BaseDTO):
    type: AuctionClientEventType
    payload: Any | None


class AuctionServerEventEnvelopeDTO(BaseDTO):
    type: AuctionServerEventType
    payload: Any | None


class AuctionPublishEnvelopeDTO(BaseDTO):
    type: AuctionPublishType
    payload: Any | None


class LeaderConnectedRequestPayloadDTO(BaseDTO):
    leader_id: int


class LeaderDisconnectedRequestPayloadDTO(BaseDTO):
    leader_id: int


class CreateRequestPayloadDTO(BaseDTO):
    auction_id: BigInt
    preset_snapshot: PresetDetailDTO


class BidErrorResponsePayloadDTO(BaseDTO):
    leader_id: int
    code: int


class AuctionRequestEnvelopeDTO(BaseDTO):
    type: AuctionRequestType
    payload: Any | None


class AuctionResponseEnvelopeDTO(BaseDTO):
    type: AuctionResponseType
    payload: Any | None

File: vite/src/dtos/auction.ts
import { z } from "zod";
import type { PresetDetailDTO } from "./preset";

export enum Status {
  WAITING = 0,
  RUNNING = 1,
  COMPLETED = 2,
}

export interface AuctionDTO {
  auctionId: string;
  guildId: string;
  presetId: number;
  status: Status;
  currentMemberId: number | null;
  currentBid: BidDTO | null;
  timer: number;
}

export interface AuctionDetailDTO extends AuctionDTO {
  teams: TeamDTO[];
  auctionQueue: number[];
  unsoldQueue: number[];
  connectedMemberIds: number[];
  presetSnapshot: PresetDetailDTO | null;
}

export interface InitPayloadDTO {
  auction: AuctionDetailDTO;
  teamId: number | null;
  memberId: number | null;
}

export interface CreateAuctionDTO {
  isPublic: boolean;
  sendInvite: boolean;
}

export enum AuctionMessageType {
  AUTH = 0,
  INIT = 1,
  ERROR = 2,
  TIMER = 3,
  STATUS = 4,
  PLACE_BID = 5,
  BID_PLACED = 6,
  MEMBER_SOLD = 7,
  MEMBER_UNSOLD = 8,
  MEMBER_CONNECTED = 9,
  MEMBER_DISCONNECTED = 10,
  NEXT_MEMBER = 11,
}

export interface AuctionMessageEnvelopeDTO<TPayload = unknown> {
  type: AuctionMessageType;
  payload: TPayload;
}

export interface TeamDTO {
  teamId: number;
  leaderId: number;
  memberIds: number[];
  points: number;
}

export const TeamSchema = z.object({
  teamId: z.number().int(),
  leaderId: z.number().int(),
  memberIds: z.array(z.number().int()),
  points: z.number().int(),
});

export interface BidDTO {
  amount: number;
  leaderId: number;
}

export interface StatusPayloadDTO {
  status: Status;
}

export interface BidPlacedPayloadDTO {
  leaderId: number;
  amount: number;
}

export interface NextMemberPayloadDTO {
  memberId: number;
  auctionQueue: number[];
  unsoldQueue: number[];
}

export interface MemberSoldPayloadDTO {
  teams: TeamDTO[];
  auctionQueue: number[];
  unsoldQueue: number[];
}

export interface MemberUnsoldPayloadDTO {
  memberId: number;
}

export interface MemberConnectedPayloadDTO {
  memberId: number;
}

export interface MemberDisconnectedPayloadDTO {
  memberId: number;
}

export interface ErrorPayloadDTO {
  code: number;
}

export interface TimerPayloadDTO {
  timer: number;
}

export interface AuthPayloadDTO {
  access_token: string | null;
}

export const AuthPayloadSchema = z.object({
  access_token: z.string().nullable(),
});

export const PlaceBidPayloadSchema = z.object({
  amount: z.number().int().min(1).max(10000),
});
